### PR TITLE
Rename __reserved to __clr_reserved to avoid a conflict on Android

### PR DESCRIPTION
**Pull request by qmfrederik**
On Android, the system headers define various fields named `__reserved`. This conflicts with the `#define __reserved` in some of the CLR headers.

Examples include:
* [wchar.h](https://android.googlesource.com/platform/development/+/master/ndk/platforms/android-21/include/wchar.h)
* [pthread.h](https://android.googlesource.com/platform/development/+/master/ndk/platforms/android-21/include/pthread.h)

This PR renames the CLR `__reserved` to `__clr_reserved` to avoid that conflict.

### CODE REVIEW

**Comment by qmfrederik** [comment]
@janvorli Happy to check whether there are less-instrusive ways to make this work, but I'm not certain.

FYI, also including some samples of compiler errors. Various of the NDK headers seem to use the `__reserved` name for fields, so it doesn't look like there's an easy way around this.

```
In file included from /home/fcarlier/git/coreclr/src/ToolBox/SOS/Strike/disasm.cpp:11:
In file included from /home/fcarlier/git/coreclr/src/ToolBox/SOS/Strike/strike.h:57:
/home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//sysroot/usr/include/wchar.h:46:18: error: expected member name or ';' after declaration specifiers
  char __reserved[4];
  ~~~~           ^
```

```
In file included from /home/fcarlier/git/coreclr/src/ToolBox/SOS/Strike/disasm.cpp:11:
In file included from /home/fcarlier/git/coreclr/src/ToolBox/SOS/Strike/strike.h:109:
In file included from /home/fcarlier/git/coreclr/src/ToolBox/SOS/Strike/exts.h:81:
In file included from /home/fcarlier/git/coreclr/src/ToolBox/SOS/Strike/util.h:2866:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../include/c++/4.9.x/unordered_map:41:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../include/c++/4.9.x/tuple:39:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../include/c++/4.9.x/array:38:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../include/c++/4.9.x/stdexcept:39:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../include/c++/4.9.x/string:52:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../include/c++/4.9.x/bits/basic_string.h:39:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../include/c++/4.9.x/ext/atomicity.h:35:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../include/c++/4.9.x/aarch64-linux-android/bits/gthr.h:148:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../include/c++/4.9.x/aarch64-linux-android/bits/gthr-default.h:35:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//sysroot/usr/include/pthread.h:33:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//sysroot/usr/include/signal.h:37:
/home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//sysroot/usr/include/asm/sigcontext.h:30:24: error: expected member name or ';' after declaration specifiers
 __u8 __reserved[4096] __attribute__((__aligned__(16)));
 ~~~~                  ^
```

and

```
In file included from /home/fcarlier/git/coreclr/src/ToolBox/SOS/Strike/disasm.cpp:11:
In file included from /home/fcarlier/git/coreclr/src/ToolBox/SOS/Strike/strike.h:109:
In file included from /home/fcarlier/git/coreclr/src/ToolBox/SOS/Strike/exts.h:81:
In file included from /home/fcarlier/git/coreclr/src/ToolBox/SOS/Strike/util.h:2866:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../include/c++/4.9.x/unordered_map:41:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../include/c++/4.9.x/tuple:39:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../include/c++/4.9.x/array:38:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../include/c++/4.9.x/stdexcept:39:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../include/c++/4.9.x/string:52:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../include/c++/4.9.x/bits/basic_string.h:39:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../include/c++/4.9.x/ext/atomicity.h:35:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../include/c++/4.9.x/aarch64-linux-android/bits/gthr.h:148:
In file included from /home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//bin/../lib/gcc/aarch64-linux-android/4.9.x/../../../../include/c++/4.9.x/aarch64-linux-android/bits/gthr-default.h:35:
/home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//sysroot/usr/include/pthread.h:47:18: error: expected member name or ';' after declaration specifiers
  char __reserved[36];
  ~~~~           ^
/home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//sysroot/usr/include/pthread.h:73:18: error: expected member name or ';' after declaration specifiers
  char __reserved[44];
  ~~~~           ^
/home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//sysroot/usr/include/pthread.h:87:18: error: expected member name or ';' after declaration specifiers
  char __reserved[16];
  ~~~~           ^
/home/fcarlier/git/coreclr/cross/android-rootfs/toolchain/arm64/sysroot/..//sysroot/usr/include/pthread.h:107:18: error: expected member name or ';' after declaration specifiers
  char __reserved[36];
  ~~~~           ^

```

**Comment by janvorli** [comment]
@qmfrederik as you can see, all the windows builds are failing. The reason is that none of the files from src/pal are included on windows, so the __clr_reserved is not defined on windows. The SAL definitions come from the Windows SDK headers when building on Windows. 
It seems that putting `#define __clr_reserved __reserved` to palclr.h by the end could fix that.

**Comment by qmfrederik** [comment]
@janvorli That didn't seem to work, let me have another look at this later today.

**Comment by qmfrederik** [comment]
@janvorli The Windows build passes now, PTAL.

File: src/pal/inc/rt/specstrings_strict.h [hunk 1]
```diff
@@ -626,7 +626,7 @@
 #define __in_awcount(expr,size)  __allowed(on_parameter)   
 #define __nullterminated         _SAL_VERSION_CHECK(__nullterminated)
 #define __nullnullterminated     _SAL_VERSION_CHECK(__nullnullterminated)
-#define __reserved               _SAL_VERSION_CHECK(__reserved)
+#define __clr_reserved               _SAL_VERSION_CHECK(__reserved)
```
Comment (janvorli) [inline src/pal/inc/rt/specstrings_strict.h]: Just a nit here and above - the indentation is wrong now. Otherwise LGTM.

Comment (qmfrederik) [inline src/pal/inc/rt/specstrings_strict.h]: OK, fixed it.

File: src/pal/inc/rt/sal.h [hunk 2]
```diff
@@ -2861,7 +2861,7 @@ of each annotation, see the advanced annotations section.
 #define __success(expr)                      _Success_(expr)
 #define __nullterminated                     _Null_terminated_
 #define __nullnullterminated
-#define __reserved                           _SAL1_Source_(__reserved, (), _Reserved_)
+#define __clr_reserved                           _SAL1_Source_(__reserved, (), _Reserved_)
```
Comment (janvorli) [inline src/pal/inc/rt/sal.h]: It is still not aligned here (that's what I meant by "the other place above")

**Comment by qmfrederik** [comment]
@janvorli Fix it, plus another place where spacing was out of line. Hope I got it correct this time!

**Pull request merged**